### PR TITLE
use gcc/g++ for SITL build on mac osx

### PR DESCRIPTION
**Pull request by bugobliterator**
the default clang produces a build that crashes at startup. This patch forces user to use gcc on their mac system for build else throws an error.

Steps for getting gcc over mac (ensure brew is already installed):
`brew install gcc`
check location of installation using `brew list gcc`
```
echo 'export PATH=/usr/local/Cellar/gcc/7.2.0_1/bin:$PATH' >> ~/.bash_profile
cp /usr/local/Cellar/gcc/7.2.0_1/bin/gcc-7 /usr/local/Cellar/gcc/7.2.0_1/bin/gcc
cp /usr/local/Cellar/gcc/7.2.0_1/bin/g++-7 /usr/local/Cellar/gcc/7.2.0_1/bin/g++
source ~/.bash_profile
```
Note: Replace `/usr/local/Cellar/gcc/7.2.0_1/bin/gcc-7` with the one you got from `brew list gcc`

### CODE REVIEW

**Comment by peterbarker** [comment]
LGTM

File: Tools/ardupilotwaf/toolchain.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ def configure(cfg):
 from waflib.Configure import conf
 from waflib.Tools import compiler_c, compiler_cxx
 from waflib.Tools import clang, clangxx, gcc, gxx
-
+from waflib import Logs
 import os
```
Comment (OXINARF) [inline Tools/ardupilotwaf/toolchain.py]: Please leave the empty line there, it's nice to separate library modules from "system" modules.

File: Tools/ardupilotwaf/toolchain.py [hunk 2]
```diff
@@ -126,13 +126,17 @@ def find_toolchain_program(cfg, filename, **kw):
     return cfg.find_program(filename, **kw)
 
 def configure(cfg):
+    _filter_supported_c_compilers('gcc', 'clang')
+    _filter_supported_cxx_compilers('g++', 'clang++')
+    
     if cfg.env.TOOLCHAIN == 'native':
         cfg.load('compiler_cxx compiler_c')
+        if cfg.env.COMPILER_CC == 'clang' or cfg.env.COMPILER_CXX == 'clang++':
+            if cfg.env.DEST_OS == 'darwin':
```
Comment (OXINARF) [inline Tools/ardupilotwaf/toolchain.py]: No need to to check for macOS, clang isn't being looked after at this point, giving the warning in all systems is good.
Also, we should put the warning for all builds, so before the _native_ check.

File: Tools/ardupilotwaf/toolchain.py [hunk 3]
```diff
@@ -126,13 +126,17 @@ def find_toolchain_program(cfg, filename, **kw):
...
+                Logs.pprint('RED', "Warning! native clang build could be unstable on macOS, Please use GCC/G++. \
```
Comment (OXINARF) [inline Tools/ardupilotwaf/toolchain.py]: Any reason not to use `Logs.warn`?

Comment (bugobliterator) [reply Tools/ardupilotwaf/toolchain.py]: do you know of alternative?

File: Tools/ardupilotwaf/toolchain.py (hunk 2 above)
Comment (bugobliterator) [reply Tools/ardupilotwaf/toolchain.py]: ok

**Comment by khancyr** [comment]
LGTM

File: Tools/ardupilotwaf/toolchain.py (hunk 3 above)
Comment (OXINARF) [reply Tools/ardupilotwaf/toolchain.py]: `Logs.warn("message")` is the alternative 🙂 By the way, I'm not sure that `\n` is handled correctly in Windows, but we can fix that later if it isn't.

**Comment by bugobliterator** [comment]
@OXINARF fixed

**Comment by OXINARF** [comment]
@bugobliterator You forgot about:

> Also, we should put the warning for all builds, so before the _native_ check.

**Comment by bugobliterator** [comment]
@OXINARF it doesn't know about the compiler before the call to cfg.load, do you want me to add separately for non native builds as well?

**Comment by OXINARF** [comment]
@bugobliterator Ah right. Leave as is then, I have some patches to remove the AR detection from there so I can change this later.

**Comment by tkoch96** [comment]
Hi, trying to run sim_vehicle.py (ArduCopter sitl) and getting a floating point error. 

I ran the commands above but sim_vehicle.py still says 'Checking for clang++' : /usr/bin/clang++.

Could I have pointers on how to get the simulator working?

Putting a full error message below:


SIM_VEHICLE: Start
SIM_VEHICLE: Killing tasks
SIM_VEHICLE: Starting up at -35.363261,149.165230,584,353 (CMAC)
SIM_VEHICLE: WAF build
SIM_VEHICLE: Configure waf
SIM_VEHICLE: "/Users/clairekolensky/Downloads/px4/ardupilot/modules/waf/waf-light" "configure" "--board" "sitl"
Setting top to                           : /Users/clairekolensky/Downloads/px4/ardupilot 
Setting out to                           : /Users/clairekolensky/Downloads/px4/ardupilot/build 
Autoconfiguration                        : enabled 
Setting board to                         : sitl 
Checking for 'clang++' (C++ compiler)    : /usr/bin/clang++ 
Checking for 'clang' (C compiler)        : /usr/bin/clang 
Checking for HAVE_CMATH_ISFINITE         : yes 
Checking for HAVE_CMATH_ISINF            : yes 
Checking for HAVE_CMATH_ISNAN            : yes 
Checking for NEED_CMATH_ISFINITE_STD_NAMESPACE : yes 
Checking for NEED_CMATH_ISINF_STD_NAMESPACE    : yes 
Checking for NEED_CMATH_ISNAN_STD_NAMESPACE    : yes 
Checking for header endian.h                   : not found 
Checking for header byteswap.h                 : not found 
Checking for program 'python'                  : /usr/bin/python 
Checking for python version >= 2.7.0           : 2.7.10 
Checking for program 'python'                  : /usr/bin/python 
Checking for python version >= 2.7.0           : 2.7.10 
Source is git repository                       : yes 
Update submodules                              : yes 
Checking for program 'git'                     : /usr/bin/git 
Checking for program 'size'                    : /usr/bin/size 
Benchmarks                                     : disabled 
Unit tests                                     : enabled 
Checking for program 'rsync'                   : /usr/bin/rsync 
'configure' finished successfully (1.216s)
SIM_VEHICLE: Building
SIM_VEHICLE: "/Users/clairekolensky/Downloads/px4/ardupilot/modules/waf/waf-light" "build" "--target" "bin/arducopter"
Waf: Entering directory `/Users/clairekolensky/Downloads/px4/ardupilot/build/sitl'
Waf: Leaving directory `/Users/clairekolensky/Downloads/px4/ardupilot/build/sitl'

BUILD SUMMARY
Build directory: /Users/clairekolensky/Downloads/px4/ardupilot/build/sitl
Target          Text     Data    BSS  Total     
------------------------------------------------
bin/arducopter  1150976  106496    0  4295368704

Build commands will be stored in build/sitl/compile_commands.json
'build' finished successfully (3.188s)
SIM_VEHICLE: Using defaults from (/Users/clairekolensky/Downloads/px4/ardupilot/Tools/autotest/default_params/copter.parm)
SIM_VEHICLE: Run ArduCopter
SIM_VEHICLE: "/Users/clairekolensky/Downloads/px4/ardupilot/Tools/autotest/run_in_terminal_window.sh" "ArduCopter" "/Users/clairekolensky/Downloads/px4/ardupilot/build/sitl/bin/arducopter" "-S" "-I0" "--home" "-35.363261,149.165230,584,353" "-w" "--model" "+" "--speedup" "1" "--defaults" "/Users/clairekolensky/Downloads/px4/ardupilot/Tools/autotest/default_params/copter.parm"
SIM_VEHICLE: Run MavProxy
SIM_VEHICLE: "mavproxy.py" "--master" "tcp:127.0.0.1:5760" "--sitl" "127.0.0.1:5501" "--out" "127.0.0.1:14550" "--out" "127.0.0.1:14551"
Connect tcp:127.0.0.1:5760 source_system=255
Failed to load module: No module named adsb. Use 'set moddebug 3' in the MAVProxy console to enable traceback
Log Directory: 
Telemetry log: mav.tlog
Waiting for heartbeat from tcp:127.0.0.1:5760
 MAV> 

Init APM:Copter V3.6-dev (0626d105)

Free RAM: 131072
FW Ver: 120
----------------------------------------


Firmware change: erasing EEPROM...
done.
load_all took 0us
0 0 0 validate_structures:304: Validating structures
DataFlash_File: buffer size=16384
online system 1
STABILIZE> Mode STABILIZE
APM: Calibrating barometer
APM: APM:Copter V3.6-dev (0626d105)
APM: Frame: QUAD
Received 886 parameters
Saved 886 parameters to mav.parm
APM: Barometer calibration complete
Init Gyro***no link
link 1 down
no link

**Comment by bugobliterator** [comment]
@tkoch96 try the with the patch in this PR and follow the steps mentioned under the initial comment.

@peterbarker should this be pulled in?

**Comment by tkoch96** [comment]
Thanks, I edited the files myself and now I don't get the error.

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**